File: python/2020/day_08.py
from pathlib import Path


def fix(program: list) -> int:
    swap = {'nop': 'jmp', 'jmp': 'nop'}
    for i, (op, _) in enumerate(program):
        if op in swap:
            op = program[i][0] = swap[op]
            if (v := run(program)) is not None:
                return v
            program[i][0] = swap[op]


def run(program: list, acc_on_loop=False) -> int:
    visited = [False] * len(program)
    i = acc = 0
    while i < len(program):
        if visited[i]:
            return acc if acc_on_loop else None
        visited[i] = True
        op, n = program[i]
        if op == 'acc':
            acc += n
        elif op == 'jmp':
            i += n - 1
        i += 1
    return acc


def main():
    text = (Path(__file__).parent / "../../input/2020/input_08.txt").read_text()
    program = []
    for line in text.strip().splitlines():
        op, n = line.split(' ')
        program.append([op, int(n)])
    print("Part 1:", run(program, acc_on_loop=True))
    print("Part 2:", fix(program))


if __name__ == "__main__":
    main()

File: rust/2020/src/bin/day_08.rs
use std::fs;

fn swap(op: &mut &str) {
    match *op {
        "nop" => *op = "jmp",
        "jmp" => *op = "nop",
        _ => (),
    }
}

fn fix(program: &mut [(&str, i32)]) -> i32 {
    for i in 0..program.len() {
        swap(&mut program[i].0);
        if let Some(v) = run(program, false) {
            return v;
        }
        swap(&mut program[i].0);
    }
    unreachable!()
}

fn run(program: &[(&str, i32)], acc_on_loop: bool) -> Option<i32> {
    let mut visited = vec![false; program.len()];
    let (mut i, mut acc) = (0, 0);
    while i < program.len() {
        if visited[i] {
            return match acc_on_loop {
                true => Some(acc),
                false => None,
            };
        }
        visited[i] = true;
        let (op, n) = program[i];
        match op {
            "acc" => acc += n,
            "jmp" => i = (i as i32 + n - 1) as usize,
            _ => (),
        }
        i += 1;
    }
    Some(acc)
}

fn main() {
    let input = fs::read_to_string("../input/2020/input_08.txt").unwrap();
    let mut program: Vec<_> = input
        .lines()
        .map(|line| {
            let tokens: Vec<_> = line.split(' ').collect();
            (tokens[0], tokens[1].parse::<i32>().unwrap())
        })
        .collect();
    println!("Part 1: {}", run(&program, true).unwrap());
    println!("Part 2: {}", fix(&mut program));
}
